pydicom/framereader.py

```diff
@@ -0,0 +1,1182 @@
+            )
+        except Exception as exc:
+            logger.error(
+                "Exception raised when attempting to parse dataset",
+                exc_info=True
+            )
+            raise IOError(
+                f"Failed to read dataset for file due to exception ({exc}). "
+                "Cannot read frames for dataset."
+            )
+        if not stopped_at_pixel_data:
+            raise IOError("Cannot read frames for dataset - missing PixelData")
+        return dataset
+
+    def to_info_dict(self) -> dict:
+        """create a dictionary representation of the FrameDataset instance for
+        convenient storage/caching
+
+        Returns
+        -------
+        dict
+            dictionary with keys `dicom_json`, `is_little_endian`,
+            `is_implicit_VR` and `TransferSyntaxUID`
+        """
+        copy_attributes = _OPTIONAL_DATASET_ATTRIBUTES + ("NumberOfFrames",)
+        ds_copy = get_dataset_copy_with_frame_attrs(self, copy_attributes)
+        return {
+            "dicom_json": ds_copy.to_json_dict(),
+            "is_little_endian": ds_copy.is_little_endian,
+            "is_implicit_VR": ds_copy.is_implicit_VR,
+            "TransferSyntaxUID": self.file_meta.TransferSyntaxUID,
+        }
+
+    @classmethod
+    def from_file(
+        cls,
+        fp: BinaryIO,
+        defer_size: Optional[Union[int, str, float]] = None,
+        force: bool = False,
+        specific_tags: Optional[List[BaseTag]] = None,
+    ) -> "FrameDataset":
+        """instantiate `FrameDataset` from a file-like object
+
+        Parameters
+        ----------
+        fp: DicomFileLike or BinaryIO
+            the file-like object for which to instantiate FrameInfo
+        defer_size: int, str or float, optional
+            See :func:`dcmread` for parameter info.
+        force: bool
+            See :func:`dcmread` for parameter info.
+        specific_tags: list or None
+            See :func:`dcmread` for parameter info.
+
+        Returns
+        -------
+        FrameDataset
+
+        """
+        dataset = cls.read_dataset(
+            file_obj=fp,
+            defer_size=defer_size,
+            force=force,
+            specific_tags=specific_tags,
+        )
+        return cls(dataset)
+
+    @classmethod
+    def from_info_dict(cls, info_dict: dict) -> "FrameDataset":
+        """Instantiate class from info_dict generated by self.to_info_dict"""
+        new_dataset = Dataset.from_json(info_dict["dicom_json"])
+        for attr in ("is_little_endian", "is_implicit_VR"):
+            setattr(new_dataset, attr, info_dict[attr])
+        if not hasattr(new_dataset, "file_meta"):
+            new_dataset.fix_meta_info(enforce_standard=False)
+        setattr(
+            new_dataset.file_meta,
+            "TransferSyntaxUID",
+            info_dict["TransferSyntaxUID"]
+        )
+        frame_dataset = cls(new_dataset)
+        return frame_dataset
+
+
+class FrameInfo:
```

Comment:
Also, be aware that there are datasets with hundreds of thousands of frames and we don't want this class to become a performance bottleneck.

Reply:
I don't see any way that this class imposes a performance bottleneck that ImageFileReader does not.